util/topgen/merge.py

```diff
@@ -408,18 +408,35 @@ def xbar_cross_node(node_name, device_xbar, xbars, visited=[]):
 
     visited.pop()
 
-    # TODO: simplify the result? if contiguous, combine into one?
-    return simplify_addr(result, device_xbar)
+    return result
+
+
+def simplify_xbar(xbars):
+    """ Simplify the address of xbar device port.
+
+    It cannot be done without fully expanded addresses.
+    So, this should be done after adding all xbar addresses.
+    """
+
+    for xbar in xbars:
+        xdev_list = [
+            x for x in xbar["nodes"] if x["type"] == "device" and x["xbar"]
+        ]
 
+        for e in xdev_list:
+            e["addr_range"] = simplify_addr(e, xbar)
 
-def simplify_addr(addrs, xbar):
+
+# TODO: Move to tlgen
+def simplify_addr(dev, xbar):
     """If any contiguous regions exist, concatenate them
 
     For instance, 0x1000 ~ 0x1FFF , 0x2000~ 0x2FFF ==> 0x1000 ~ 0x2FFF
 
```

Comment:
I would recommend enhancing the comments here also to talk about non-continuguous, other devices blocking and non-blocking.